certora/spec/KintoID/SanctionsTraits.spec

```diff
@@ -0,0 +1,397 @@
+    if(companyOrIndividual) {
+        mintCompanyKyc(e, signatureData, traits);
+    }
+    else {
+        mintIndividualKyc(e, signatureData, traits);
+    }
+        
+    address ownerAfter = ownerOf(tokenID);
+
+    assert ownerBefore != ownerAfter;
+    assert ownerAfter == account;
+}
+
+rule onlyKYCCanChangeBalance(address account, method f) filtered{f -> !viewOrUpgrade(f)} {
+    uint256 balanceBefore = balanceOf(account);
+        env e;
+        calldataarg args;
+            f(e,args);
+    uint256 balanceAfter = balanceOf(account);
+
+    assert balanceBefore != balanceAfter => hasRole(KYC_PROVIDER_ROLE(), e.msg.sender);
+}
+
+rule burnByOwnerLiveness(uint256 tokenID) {
+    address owner = ownerOf(tokenID);
+
+    require balanceOf(owner) != 0; /// Can only burn if already minted (thus has balance).
+    require totalSupply() != 0; /// Assuming totalSupply() == sum of balances.
+    uint256 tokenIDEnd;
+    require tokensIndex[tokenIDEnd] == assert_uint256(NumberOfTokens - 1);
+    requireInvariant TokenIndexIsUpToArrayLength(tokenID);
+    requireInvariant TokenIndexIsUpToArrayLength(tokenIDEnd);
+    env e; 
+        require e.msg.sender == owner; /// Owner of token is the msg.sender.
+        require e.msg.value == 0; /// Not a payable function.
+    burn@withrevert(e, tokenID);
+    
+    assert !lastReverted;
+}
+
+rule cannotBurnRightAfterMint(IKintoID.SignatureData signatureData) {
+    bool companyOrIndividual;
+    env e1;
+    env e2;
+    uint8[] traits;
+    if(companyOrIndividual) {
+        mintCompanyKyc(e1, signatureData, traits);
+    }
+    else {
+        mintIndividualKyc(e1, signatureData, traits);
+    }
+    burnKYC@withrevert(e2, signatureData);
+
+    assert lastReverted;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: sanctions                                                                                                    │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+invariant hasSanctionCountIsNonZero(env e1, address account, uint16 CID)
+    isSanctioned(e1, account, CID) => getSanctionsCount(account) > 0
+    filtered{f -> !upgradeMethods(f)}
+    {
+        preserved with (env e2) {
+            require e2.block.timestamp == e1.block.timestamp;
+        }
+        preserved monitor(address[] accounts, IKintoID.MonitorUpdateData[][] data) with (env e2) {
+            require e2.block.timestamp == e1.block.timestamp;
+            if(accounts.length > 0) {
+                require accounts.length == 1;
+                require data[0].length == 1;
+                uint16 CID_A = data[0][0].index;
+                address accountA = accounts[0];
+                require (accountA == account && 
+                isSanctioned(e2, accountA, CID_A) && isSanctioned(e2, accountA, CID)) => getSanctionsCount(accountA) > 1;
```

Comment:
isn't it better to just correlate: CID_A == CID and then no need to repeat the invariant ?

Reply:
What do you mean? The CIDs can be different.